Introduce framework for filtering/selecting pluggable components for sonic modular chassis

rawal01 opened the pull request:
Introduce framework for filtering/selecting pluggable components for sonic modular chassis

* Added new script:  `tests/platform_tests/cli/test_show_chassis_module.py`
  * Added two new test cases for verifying `show chassis-module status` and `show chassis-module midplane status` that were introduced as part of PRs https://github.com/Azure/sonic-utilities/pull/1145/ and https://github.com/Azure/sonic-utilities/pull/1267
* Updated `tests/platform_tests/conftest.py`
   * Added a fixture that skips checking for modules for DUT

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ x] Test case(new/improvement)

### Approach
#### What is the motivation for this PR?
Add test cases to verify show chassis-module status and show chassis-module midplane-status for VOQ chassis. The cli commands were introduced in following PRs:
* https://github.com/Azure/sonic-utilities/pull/1145/ 
* https://github.com/Azure/sonic-utilities/pull/1267

#### How did you do it?

1. Added new script `tests/platform_tests/cli/test_show_chassis_module.py` for verifying show chassis-module status and show chassis-module midplane status. Introduced test cases:
  * `test_show_chassis_module_status`: verify the output of  command show chassis-module status
  * `test_show_chassis_module_midplane_status`: verify the output of  command show chassis-module midplane-status
2. Added new script `test_power_budget_info.py` for verifying the redis output for power budget policy in supervisor card of chassis
3. Added a fixture that skips checking for modules for DUT. This new fixture is introduced to skips modules for DUT not present based on entries in inventory file, by default the test it is assumed the chassis is fully equipped and all modules are up. Example to skip certain modules add to skip modules in inventory for DUT:
```
DUT1
    skip_modules:
        'line-cards':
          - LINE-CARD0
          - LINE-CARD2
        'fabric-cards':
          - FABRIC-CARD3
        'psus':
          - PSU4
          - PSU5
```
Based on inventory file the tests will skip above modules for DUT1. For example, `show chassis module` will allow empty as status for LINE-CARD0 and LINE-CARD2 while all other will be expected to be ONLINE.
 
#### How did you verify/test it?
Ran sample  tests against Nokia chassis with t2 topology using the skip fixture

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?
t2

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

jleveque commented on ansible/platform_skip_inventory.json:
```diff
@@ -0,0 +1,9 @@
+{
+   "dut2": {
+       "modules": {
+                "skip": ["FABRIC-CARD0", "LINE-CARD4"]
+           }
+      }
```
If you want to determine which modules to test or skip on a DuT basis, it might be better to add attributes to the inventory file rather than adding a new JSON file.

rawal01 commented:
retest vsimage please

wangxin commented on tests/platform_tests/cli/test_show_chassis_module.py:
```diff
@@ -0,0 +1,82 @@
+
+def test_show_chassis_module_status(duthosts, enum_dut_hostname):
+    cmd = " ".join([CMD_SHOW_CHASSIS_MODULE, "status"])
+    logging.info("verifying output of cli command {}".format(cmd))
```
It would be better to initialize a logger than directly using the logging module.

rawal01 replied:
@wangxin 
I made changes and committed them as per your comment

wangxin commented on tests/platform_tests/cli/test_show_chassis_module.py:
```diff
@@ -0,0 +1,82 @@
+CMD_SHOW_CHASSIS_MODULE = "show chassis-module"
+
+
+def parse_chassis_module(output, expected_headers):
```
If the output is of tabulate format, you can consider using this method to parse the output: https://github.com/Azure/sonic-mgmt/blob/master/tests/common/devices.py#L1034

rawal01 replied:
@wangxin I need to verify headers for show command which is done in parse function which is already using get_field_range, get_fields utils from utils under cli folder

wangxin reviewed (commented):
Can you address the LGTM alerts?

wangxin commented:
@rawal01 Can you update the PR description to match the latest code of this PR?

saravanansv commented on tests/platform_tests/test_power_budget_info.py:
```diff
@@ -0,0 +1,60 @@
+                n_psu = (re.split('Supplied Power ', out_val))[1]
+                if n_psu not in skip_mod_list:
+                    sup_power = float(out_dict[pb_name]['value'][out_val])
+                    pytest_assert(sup_power > 0,
+                                  "expected supplied power for psu {} is expected to 0 or less".format(n_psu))
```
psu {} is expected to be > 0 ? 
And can it print the sup_power value as well ?

rawal01 replied:
not psu {} but supplied power for psu expected to be greater than 0

saravanansv commented on tests/platform_tests/test_power_budget_info.py:
```diff
@@ -0,0 +1,60 @@
+        tot_supp_power = float(out_dict[pb_name]['value']['Total Supplied Power'])
+        tot_cons_power = float(out_dict[pb_name]['value']['Total Consumed Power'])
+        pytest_assert(exp_total_cons_power == tot_cons_power,
+                      "total consumed power is not correct expected is {} reported is {}".format(
```
"not correct" -> "incorrect"

rawal01 replied:
@saravanansv saravanansv changed as per your comment

saravanansv commented on tests/platform_tests/cli/test_show_chassis_module.py:
```diff
@@ -0,0 +1,83 @@
+        pytest_assert(header_v in headers, "Missing header {}".format(header_v))
+
+    result = {}
+    for a_line in output[2:]:
```
a_line can be renamed as just line

rawal01 commented on tests/platform_tests/cli/util.py:
```diff
@@ -62,3 +62,38 @@ def get_fields(line, field_ranges):
+    if 'skip_modules' in dut_vars:
+        if mod_key is None:
+            for k in dut_vars['skip_modules'].keys():
+                for i in dut_vars['skip_modules'][k]:
```
@saravanansv  I changed to name vars more user friendly please review the changes

rawal01 commented:
> @rawal01 Can you update the PR description to match the latest code of this PR?
@wangxin
updated description to remove reference to JSON file and updated fixture description

wangxin commented:
> > @rawal01 Can you update the PR description to match the latest code of this PR?
> > @wangxin
> > updated description to remove reference to JSON file and updated fixture description

I have reformatted and updated your PR description a little bit. Could you check if it is accurate?

rawal01 commented:
> > > @rawal01 Can you update the PR description to match the latest code of this PR?
> > > @wangxin
> > > updated description to remove reference to JSON file and updated fixture description
> 
> I have reformatted and updated your PR description a little bit. Could you check if it is accurate?

Looks good

shubav commented:
@wangxin @yxieca could we merge this PR? We have new tests for the new chassis APIs and have converted existing API tests to work for chassis. We can push that once this is merged, as there's a dependency. Thanks.
